internal/dataplane/parser/golden_test.go

```diff
@@ -0,0 +1,324 @@
+package parser_test
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/blang/semver/v4"
+	"github.com/samber/lo"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"k8s.io/kubectl/pkg/cmd/util"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/deckgen"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/dataplane/parser"
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/store"
+)
+
+var (
+	// updateGolden tells whether to update golden files using the current output of the parser.
+	updateGolden = os.Getenv("UPDATE_GOLDEN") == "true"
+
+	// defaultKongVersion is the default Kong version to use in tests. Can be overridden in a test case.
+	defaultKongVersion = semver.MustParse("3.3.0")
+
+	// defaultFeatureFlags is the default set of feature flags to use in tests. Can be overridden in a test case.
+	defaultFeatureFlags = func() parser.FeatureFlags {
+		return parser.FeatureFlags{
+			CombinedServiceRoutes: true,
+			RegexPathPrefix:       true,
+		}
+	}
+)
+
+// TestParser_GoldenTests runs the golden tests for the parser.
+//
+// Command to update the golden files:
+// $ make test.golden.update
+//
+// Data for the test cases is stored in the ./testdata/golden directory. It's recommended to group test cases into
+// subdirectories based on the Kubernetes input that they run against so that each of the subdirectories has:
+//   - an input file that represents the input with Kubernetes objects to be loaded into the store (in.yaml),
+//   - a set of expected golden *.yaml files (in Deck format) where each file represents an expected output for a given
+//     parser configuration (feature flags, Kong version, etc.).
+//
+// The test case is executed by loading the in.yaml file into the store, then running the parser on the store,
+// and finally comparing the output of the parser with the expected golden file.
+//
+// When adding a new test case, you can follow these steps:
+// - Add a test case to the testCases slice below along with.
+// - Add a new directory ./testdata/golden/<your-dir> with the in.yaml that you want to test against.
+// - Specify the path for the golden file (./testdata/golden/<your-dir>/<your-parser-config>.yaml).
+// - Run `make test.golden.update` to generate the golden file.
+// - Inspect the generated golden file and make sure it's correct.
+//
+// If you introduce a change that may affect many test cases, and you're sure about it correctness, you can run the
+// command as well to update all golden files at once.
+func TestParser_GoldenTests(t *testing.T) {
+	const (
```

Comment:
Just to consider:

Since human beings are prone to make errors or forget things, I suggest to consider making these testcases self contained in one place, i.e. to put all the test data test metadata in the directories for said test cases.

The reason is that we might forget adding tests here, thinking that the test machinery would pick this up automatically or doing some other mistakes that might be hard to foresee now, or debug later :)

test name could be the directory name containing the test data, feature gates could be stored in a dedicated file in a [dot env file format](https://hexdocs.pm/dotenvy/dotenv-file-format.html)

Golden file could be just named `golden` (or anything else for that matter).

I wouldn't be too concerned about test file or directories duplication.

WDYT?